cmd/runtimetest/main.go

```diff
@@ -244,6 +245,45 @@ func validateROPaths(spec *rspec.Spec) error {
 	return nil
 }
 
+func getThisCgroupPath(subsystem string) (string, error) {
+	f, err := os.Open("/proc/self/cgroup")
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+
+	s := bufio.NewScanner(f)
+	for s.Scan() {
+		if err := s.Err(); err != nil {
+			return "", err
+		}
+
+		text := s.Text()
+		parts := strings.Split(text, ":")
+
+		for _, subs := range strings.Split(parts[1], ",") {
+			if subs == subsystem {
+				return parts[2], nil
+			}
+		}
+	}
+	return "", fmt.Errorf("cgroup %v not found", subsystem)
+}
+
+func validateCgroupsPath(spec *rspec.Spec) error {
+	fmt.Println("validating cgroupsPath")
+	expectedPath := spec.Linux.CgroupsPath
+	actualPath, err := getThisCgroupPath("devices")
```

Comment:
On Thu, Jun 02, 2016 at 10:31:32PM -0700, Ma Shimiao wrote:
> > +	actualPath, err := getThisCgroupPath("devices")
> 
> Well, We need many default devices to create container. So cgroup
> device should be hard requirement.

But just because you create devices doesn't mean the user is setting
anything in linux.resources.devices [1].

> The spec says "The Spec does not support split
> hierarchy[https://www.kernel.org/doc/Documentation/cgroup-v2.txt]. The
> cgroups will be created if they don't exist."  unified cgroups may
> not to be supported?

No, that wording is “we don't support per-controller paths” for the
reasons discussed in cgroup-v2.txt [2].  It's not about limiting the
cgroup version.

[1]: https://github.com/opencontainers/runtime-spec/blob/v0.5.0/config-linux.md#device-whitelist
[2]: https://github.com/opencontainers/runtime-spec/pull/137#discussion_r38601118